KCL grammar parse error

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
cd /tmp
git clone https://github.com/kcl-lang/kcl.git
cd kcl
cargo build

/tmp/kcl/_build/dist/linux/core/libkcl run test.k
error[E1001]: InvalidSyntax
 --> /tmp/kcl/test.k:5:5
  |
5 |     {
  |     ^ expected one of ["]"] got {
  |

error[E1001]: InvalidSyntax
 --> /tmp/kcl/test.k:7:5
  |
7 |     }
  |     ^ expected one of [":", "=", "+="] got newline
  |

error[E1001]: InvalidSyntax
 --> /tmp/kcl/test.k:8:5
  |
8 |     ]
  |     ^ expected one of ["identifier", "literal", "(", "[", "{"] got ]
  |

error[E1001]: InvalidSyntax
 --> /tmp/kcl/test.k:8:5
  |
8 |     ]
  |     ^ expected one of ["identifier", "literal", "(", "[", "{"] got ]
  |

error[E1001]: InvalidSyntax
 --> /tmp/kcl/test.k:8:5
  |
8 |     ]
  |     ^ expected one of [":", "=", "+="] got ]
  |

error[E1001]: InvalidSyntax
 --> /tmp/kcl/test.k:8:5
  |
8 |     ]
  |     ^ expected one of ["identifier", "literal", "(", "[", "{"] got newline
  |

error[E2A31]: IllegalAttributeError
 --> /tmp/kcl/test.k:5:5
  |
5 |     {
  |     ^ A attribute must be string type, got '{str(name):str(test2)}'
  |

```

the issue happens with the following the following kcl file:


```test.k
{
    hi: [{
        name: "test"
    }
    {
        name: "test2"
    }
    ]
}
```

but goes away if you run this version:

```working version

{
    hi: [
    {
        name: "test"
    }
    {
        name: "test2"
    }
    ]
}
```


### 4. What is your KCL components version? (Required)
Version: 0.12.3-c020ab3eb4b9179219d6837a57f5d323
Platform: x86_64-unknown-linux-gnu
GitCommit: a0dbdf6b996d6a40fb650c773a5967e4b66c7ccb

--- a/crates/parser/src/parser/expr.rs
+++ b/crates/parser/src/parser/expr.rs
@@ -964,7 +964,7 @@ impl<'a> Parser<'a> {
             ));
         }
 
-        let has_newline = if self.token.kind == TokenKind::Newline {
+        if self.token.kind == TokenKind::Newline {
             self.skip_newlines();
             self.clean_all_indentations();
             if self.token.kind == TokenKind::CloseDelim(DelimToken::Bracket) {
@@ -978,13 +978,10 @@ impl<'a> Parser<'a> {
                     self.sess.struct_token_loc(token, self.prev_token),
                 ));
             }
-            true
-        } else {
-            false
-        };
+        }
 
         let item_start_token = self.token;
-        let items = self.parse_list_items(has_newline);
+        let items = self.parse_list_items();
         let generators = self.parse_comp_clauses();
 
         // _DEDENT
@@ -1052,10 +1049,9 @@ impl<'a> Parser<'a> {
 
     /// Syntax:
     /// list_items: expr ((COMMA [NEWLINE] | NEWLINE) expr)* [COMMA] [NEWLINE]
-    pub(crate) fn parse_list_items(&mut self, has_newline: bool) -> Vec<NodeRef<Expr>> {
+    pub(crate) fn parse_list_items(&mut self) -> Vec<NodeRef<Expr>> {
         let is_terminator = |token: &kcl_ast::token::Token| match &token.kind {
             TokenKind::CloseDelim(DelimToken::Bracket) | TokenKind::Eof => true,
-            TokenKind::Newline if !has_newline => true,
             _ => token.is_keyword(kw::For),
         };
 
@@ -1068,9 +1064,7 @@ impl<'a> Parser<'a> {
         if let TokenKind::Comma = self.token.kind {
             self.bump();
         }
-        if has_newline {
-            self.skip_newlines();
-        }
+        self.skip_newlines();
         loop {
             let marker = self.mark();
             self.clean_all_indentations();
@@ -1090,9 +1084,7 @@ impl<'a> Parser<'a> {
             if let TokenKind::Comma = self.token.kind {
                 self.bump();
             }
-            if has_newline {
-                self.skip_newlines();
-            }
+            self.skip_newlines();
             self.drop(marker);
         }
         items

--- a/crates/parser/src/tests/expr.rs
+++ b/crates/parser/src/tests/expr.rs
@@ -224,3 +224,37 @@ parse_expr_snapshot!(
     a = 1
 "####
 );
+
+// Test for issue #2042: List with multiple config items without newline after [
+parse_expr_snapshot!(
+    list_with_config_items_0,
+    r####"[{
+        name: "test"
+    }
+    {
+        name: "test2"
+    }]"####
+);
+
+// Test for issue #2042: List with multiple config items with comma separators
+parse_expr_snapshot!(
+    list_with_config_items_1,
+    r####"[{
+        name: "test"
+    },
+    {
+        name: "test2"
+    }]"####
+);
+
+// Test for issue #2042: List with multiple config items with newline after [
+parse_expr_snapshot!(
+    list_with_config_items_2,
+    r####"[
+    {
+        name: "test"
+    }
+    {
+        name: "test2"
+    }]"####
+);

--- a/tests/grammar/datatype/dict/normal/main.k
+++ b/tests/grammar/datatype/dict/normal/main.k
@@ -1,0 +1,19 @@
+a = {
+    hi: [{
+        name: "test"
+    }
+    {
+        name: "test2"
+    }
+    ]
+}
+b = {
+    hi: [
+    {
+        name: "test"
+    }
+    {
+        name: "test2"
+    }
+    ]
+}

--- a/tests/grammar/datatype/dict/normal/stdout.golden
+++ b/tests/grammar/datatype/dict/normal/stdout.golden
@@ -1,0 +1,8 @@
+a:
+  hi:
+  - name: test
+  - name: test2
+b:
+  hi:
+  - name: test
+  - name: test2

